File: Data-Structures-Algorithms-In-TS/Graph/GraphAM.ts

class PriorityQueue<T> {
    static CAPACITY: number = 32;
    size: number;
    arr: Array<T>;
    compare: (a: T, b: T) => boolean;

    public constructor(cmp: (a: T, b: T) => boolean) {
        this.compare = cmp;
        this.arr = new Array<T>(PriorityQueue.CAPACITY);
        this.size = 0;
    }

    percolateDown(parent: number) {
        let lChild: number = 2 * parent + 1;
        let rChild: number = lChild + 1;
        let child: number = -1;
        let temp: T;
        if (lChild < this.size) {
            child = lChild;
        }
        if (rChild < this.size && this.compare(this.arr[lChild], this.arr[rChild])) {
            child = rChild;
        }
        if (child !== -1 && this.compare(this.arr[parent], this.arr[child])) {
            temp = this.arr[parent];
            this.arr[parent] = this.arr[child];
            this.arr[child] = temp;
            this.percolateDown(child);
        }
    }

    percolateUp(child: number) {
        let parent: number = Math.floor((child - 1) / 2);
        let temp: T;
        if (parent < 0) {
            return;
        }
        if (this.compare(this.arr[parent], this.arr[child])) {
            temp = this.arr[child];
            this.arr[child] = this.arr[parent];
            this.arr[parent] = temp;
            this.percolateUp(parent);
        }
    }

    public add(value: T) {
        if (this.size === this.arr.length) {
            this.doubleSize();
        }

        this.arr[this.size++] = value;

        this.percolateUp(this.size - 1);
    }

    private doubleSize() {
        let old: Array<T> = this.arr;
        let newSize = this.size * 2;
        if (newSize === 0)
            newSize = 1;
        this.arr = new Array<T>(newSize);
        /* arraycopy */
        let size = this.size;
        for (let i = 0; i < size; i++) {
            this.arr[i] = old[i];
        }
    }

    public remove(): T {
        if (this.isEmpty()) {
            throw new Error("IllegalStateException");
        }
        let value: T = this.arr[0];
        this.arr[0] = this.arr[this.size - 1];
        this.size--;
        this.percolateDown(0);
        return value;
    }

    public print() {
        console.log(this.arr);;
    }

    public isEmpty(): boolean {
        return (this.size === 0);
    }

    public length(): number {
        return this.size;
    }

    public peek(): T {
        if (this.isEmpty()) {
            throw new Error("IllegalStateException");
        }
        return this.arr[0];
    }
}

class Queue<T> {
    frontIndex: number;
    data: T[];

    public constructor() {
        this.frontIndex = 0;
        this.data = [];
    }

    public add(value: T) {
        this.data.push(value);
    }

    public remove(): T {
        let value = this.data[this.frontIndex];
        this.frontIndex++;
        if (this.data.length > 0 && this.frontIndex * 2 >= this.data.length) {
            this.data = this.data.slice(this.frontIndex);
            this.frontIndex = 0;
        }
        return value;
    }

    public peek(): T {
        let value = this.data[this.frontIndex];
        return value;
    }

    public isEmpty(): boolean {
        return (this.data.length - this.frontIndex) === 0;
    }

    public size(): number {
        return (this.data.length - this.frontIndex);
    }

    public peekLast(): T {
        return this.data[this.data.length - 1]
    }
}

const infi = 2147483647;

class GraphEdge {
    dest: number;
    cost: number;

    public constructor(dst: number, cst: number) {
        this.dest = dst;
        this.cost = cst;
    }
}

function EdgeComparator(x: GraphEdge, y: GraphEdge): boolean {
    return x.cost > y.cost
}

class Graph {
    count: number;
    adj: Array<Array<number>>;

    constructor(cnt: number) {
        if (cnt === undefined)
            throw new Error('Invalid argument')

        this.count = cnt;
        this.adj = new Array<Array<number>>(this.count)
        for (let i = 0; i < this.count; i++)
            this.adj[i] = new Array<number>(this.count).fill(0);
    }

    public addDirectedEdge(src: number, dst: number, cost: number = 1) {
        this.adj[src][dst] = cost;
    }

    public addUndirectedEdge(src: number, dst: number, cost: number = 1) {
        this.addDirectedEdge(src, dst, cost);
        this.addDirectedEdge(dst, src, cost);
    }

    public print() {
        for (let i: number = 0; i < this.count; i++) {
            let t = `Vertex ${i} is connected to : `;
            for (let j = 0; j < this.count; j++) {
                if (this.adj[i][j] !== 0)
                    t += j + "(cost: " + this.adj[i][j] + ") "
            }
            console.log(t)
        }
    }


    public dijkstra(source: number) {
        let previous: Array<number> = new Array<number>(this.count).fill(-1);
        let dist: Array<number> = new Array<number>(this.count).fill(infi);
        let visited: Array<boolean> = new Array<boolean>(this.count).fill(false);

        dist[source] = 0;
        previous[source] = source;

        let queue: PriorityQueue<GraphEdge> = new PriorityQueue<GraphEdge>(EdgeComparator);
        let node: GraphEdge = new GraphEdge(source, 0);
        queue.add(node);
        let curr: number;

        while (queue.isEmpty() === false) {
            node = queue.remove();
            curr = node.dest;
            visited[curr] = true;
            for (let dest: number = 0; dest < this.count; dest++) {
                let cost: number = this.adj[curr][dest];
                if (cost !== 0) {
                    let alt: number = cost + dist[curr];
                    if (dist[dest] > alt && visited[dest] === false) {
                        dist[dest] = alt;
                        previous[dest] = curr;
                        node = new GraphEdge(dest, alt);
                        queue.add(node);
                    }
                }
            };
        };
        this.printPath(previous, dist, this.count, source);
    }

    private printPathUtil(previous, source, dest): string {
        let path = "";
        if (dest === source) {
            path += source;
        } else {
            path += this.printPathUtil(previous, source, previous[dest]);
            path += "->" + dest;
        }
        return path;
    }

    private printPath(previous, dist, count, source): void {
        let output = "Shortest Paths: ";
        for (let i = 0; i < count; i++) {
            if (dist[i] === 99999) {
                output += "(" + source + "->" + i + " @ Unreachable) ";
            } else if (i !== previous[i]) {
                output += "(";
                output += this.printPathUtil(previous, source, i);
                output += " @ " + dist[i] + ") ";
            }
        }
        console.log(output);
    }


    public primsMST() {
        let previous: Array<number> = new Array<number>(this.count).fill(-1);
        let dist: Array<number> = new Array<number>(this.count).fill(0);
        let source: number = 0;
        let visited: Array<boolean> = new Array<boolean>(this.count).fill(false);
        for (let i: number = 0; i < this.count; i++) {
            previous[i] = -1;
            dist[i] = infi;
            visited[i] = false;
        };
        dist[source] = 0;
        previous[source] = -1;
        let queue: PriorityQueue<GraphEdge> = new PriorityQueue<GraphEdge>(EdgeComparator);
        let node: GraphEdge = new GraphEdge(source, 0);
        queue.add(node);
        let curr: number;

        while (queue.isEmpty() === false) {
            node = queue.remove();
            curr = node.dest;
            visited[curr] = true;

            for (let dest: number = 0; dest < this.count; dest++) {
                let cost: number = this.adj[curr][dest];
                if (cost !== 0) {
                    let alt: number = cost;
                    if (dist[dest] > alt && visited[dest] === false) {
                        dist[dest] = alt;
                        previous[dest] = curr;
                        node = new GraphEdge(dest, alt);
                        queue.add(node);
                    }
                }
            };
        };
        let total = 0;
        let output = "Edges are: ";
        for (let i = 0; i < this.count; i++) {
            if (dist[i] === Infinity) {
                output += `( ${i},  Unreachable)`
            } else if (previous[i] != i) {
                output += `(${previous[i]}->${i} @ ${dist[i]}) `
                total += dist[i];
            }
        }
        console.log(output);
        console.log(`Total MST cost: ${total}`)
    }


    private hamiltonianPathUtil(path: Array<number>, pSize: number, added: Array<number>): boolean {
        if (pSize === this.count) {
            return true;
        }
        for (let vertex: number = 0; vertex < this.count; vertex++) {
            if (pSize === 0 || (this.adj[path[pSize - 1]][vertex] === 1 && added[vertex] === 0)) {
                path[pSize++] = vertex;
                added[vertex] = 1;
                if (this.hamiltonianPathUtil(path, pSize, added)) return true;
                pSize--;
                added[vertex] = 0;
            }
        };
        return false;
    }

    public hamiltonianPath(): boolean {
        let path: Array<number> = new Array<number>(this.count).fill(0);
        let added: Array<number> = new Array<number>(this.count).fill(0);
        if (this.hamiltonianPathUtil(path, 0, added)) {
            console.log("Hamiltonian Path found :: " + path);
            return true;
        }
        console.log("Hamiltonian Path not found");
        return false;
    }

    private hamiltonianCycleUtil(path: Array<number>, pSize: number, added: Array<number>): boolean {
        if (pSize === this.count) {
            if (this.adj[path[pSize - 1]][path[0]] === 1) {
                path[pSize] = path[0];
                return true;
            } else return false;
        }
        for (let vertex: number = 0; vertex < this.count; vertex++) {
            if (pSize === 0 || (this.adj[path[pSize - 1]][vertex] === 1 && added[vertex] === 0)) {
                path[pSize++] = vertex;
                added[vertex] = 1;
                if (this.hamiltonianCycleUtil(path, pSize, added)) return true;
                pSize--;
                added[vertex] = 0;
            }
        }
        return false;
    }

    public hamiltonianCycle(): boolean {
        let path: Array<number> = new Array<number>(this.count + 1).fill(0);
        let added: Array<number> = new Array<number>(this.count).fill(0);

        if (this.hamiltonianCycleUtil(path, 0, added)) {
            console.log("Hamiltonian Cycle found :: " + path);
            return true;
        }
        console.log("Hamiltonian Cycle not found");
        return false;
    }
}

/* Testing Code */
function test1() {
    const graph = new Graph(4);
    graph.addUndirectedEdge(0, 1);
    graph.addUndirectedEdge(0, 2);
    graph.addUndirectedEdge(1, 2);
    graph.addUndirectedEdge(2, 3);
    graph.print();
}

/*
Vertex 0 is connected to : 1(cost: 1) 2(cost: 1) 
Vertex 1 is connected to : 0(cost: 1) 2(cost: 1) 
Vertex 2 is connected to : 0(cost: 1) 1(cost: 1) 3(cost: 1) 
Vertex 3 is connected to : 2(cost: 1) 
*/

/* Testing Code */
function test2() {
    const gph = new Graph(9);
    gph.addUndirectedEdge(0, 1, 4);
    gph.addUndirectedEdge(0, 7, 8);
    gph.addUndirectedEdge(1, 2, 8);
    gph.addUndirectedEdge(1, 7, 11);
    gph.addUndirectedEdge(2, 3, 7);
    gph.addUndirectedEdge(2, 8, 2);
    gph.addUndirectedEdge(2, 5, 4);
    gph.addUndirectedEdge(3, 4, 9);
    gph.addUndirectedEdge(3, 5, 14);
    gph.addUndirectedEdge(4, 5, 10);
    gph.addUndirectedEdge(5, 6, 2);
    gph.addUndirectedEdge(6, 7, 1);
    gph.addUndirectedEdge(6, 8, 6);
    gph.addUndirectedEdge(7, 8, 7);
    console.log("")
    gph.primsMST();
    console.log("")
}

/*
Edges are (0->1 @ 4) (5->2 @ 4) (2->3 @ 7) (3->4 @ 9) (6->5 @ 2) (7->6 @ 1) (0->7 @ 8) (2->8 @ 2) 
Total MST cost : 37
*/

/* Testing Code */
function test3() {
    const gph = new Graph(9);
    gph.addUndirectedEdge(0, 1, 4)
    gph.addUndirectedEdge(0, 7, 8)
    gph.addUndirectedEdge(1, 2, 8)
    gph.addUndirectedEdge(1, 7, 11)
    gph.addUndirectedEdge(2, 3, 7)
    gph.addUndirectedEdge(2, 8, 2)
    gph.addUndirectedEdge(2, 5, 4)
    gph.addUndirectedEdge(3, 4, 9)
    gph.addUndirectedEdge(3, 5, 14)
    gph.addUndirectedEdge(4, 5, 10)
    gph.addUndirectedEdge(5, 6, 2)
    gph.addUndirectedEdge(6, 7, 1)
    gph.addUndirectedEdge(6, 8, 6)
    gph.addUndirectedEdge(7, 8, 7)
    gph.dijkstra(1);
}

/*
Shortest Paths: (1->0 @ 4) (1->2 @ 8) (1->2->3 @ 15) (1->2->5->4 @ 22) (1->2->5 @ 12) (1->7->6 @ 12) (1->7 @ 11) (1->2->8 @ 10) 
*/

/* Testing Code */
function test4() {
    const count = 5;
    const graph = new Graph(count);
    const adj =
        [[0, 1, 0, 1, 0],
        [1, 0, 1, 1, 0],
        [0, 1, 0, 0, 1],
        [1, 1, 0, 0, 1],
        [0, 1, 1, 1, 0]];

    for (let i = 0; i < count; i++) {
        for (let j = 0; j < count; j++) {
            if (adj[i][j] === 1)
                graph.addDirectedEdge(i, j, 1);
        }
    }

    console.log(`hamiltonianPath : ${graph.hamiltonianPath()}`);

    const graph2 = new Graph(count);
    const adj2 =
        [[0, 1, 0, 1, 0],
        [1, 0, 1, 1, 0],
        [0, 1, 0, 0, 1],
        [1, 1, 0, 0, 0],
        [0, 1, 1, 0, 0]];

    for (let i = 0; i < count; i++) {
        for (let j = 0; j < count; j++) {
            if (adj2[i][j] === 1)
                graph2.addDirectedEdge(i, j, 1);
        }
    }
    console.log(`hamiltonianPath :  ${graph2.hamiltonianPath()}`);
}

/*
Hamiltonian Path found ::  [ 0, 1, 2, 4, 3 ]
hamiltonianPath : true

Hamiltonian Path found ::  [ 0, 3, 1, 2, 4 ]
hamiltonianPath :  true
*/

/* Testing Code */
function test5() {
    const count = 5;
    const graph = new Graph(count);
    const adj =
        [[0, 1, 0, 1, 0],
        [1, 0, 1, 1, 0],
        [0, 1, 0, 0, 1],
        [1, 1, 0, 0, 1],
        [0, 1, 1, 1, 0]];

    for (let i = 0; i < count; i++) {
        for (let j = 0; j < count; j++) {
            if (adj[i][j] === 1)
                graph.addDirectedEdge(i, j, 1);
        }
    }

    console.log(`hamiltonianCycle : ${graph.hamiltonianCycle()}`);

    const graph2 = new Graph(count);
    const adj2 =
        [[0, 1, 0, 1, 0],
        [1, 0, 1, 1, 0],
        [0, 1, 0, 0, 1],
        [1, 1, 0, 0, 0],
        [0, 1, 1, 0, 0]];

    for (let i = 0; i < count; i++) {
        for (let j = 0; j < count; j++) {
            if (adj2[i][j] === 1)
                graph2.addDirectedEdge(i, j, 1);
        }
    }
    console.log(`hamiltonianCycle :  ${graph2.hamiltonianCycle()}`);
}

/*
Hamiltonian Cycle found ::  [ 0, 1, 2, 4, 3, 0 ]
hamiltonianCycle : true

Hamiltonian Cycle not found
hamiltonianCycle :  false
*/

test1()
test2()
test3()
test4()
test5()

File: Data-Structures-and-Algorithms-using-Swift/Graph/GraphAM.swift
import Foundation


class Heap < T: Comparable > {
    private var size : Int // Number of elements in Heap
    private var arr : [T?] // The Heap array
    private var isMinHeap : Bool

    init(_ isMin : Bool) {
        self.arr = []
        self.size = 0
        self.isMinHeap = isMin
    }

    init(_ array : inout [T], _ isMin : Bool) {
        self.size = array.count
        self.arr = array
        self.isMinHeap = isMin

		// Build Heap operation over array
		var i : Int = (self.size / 2)
		while (i >= 0) {
			self.percolateDown(i)
			i -= 1
		}
    }

    func comp(_ first : T, _ second : T) -> Bool {
        if (self.isMinHeap) {
            return first > second
        } else {
            return first < second
        }
    }

    // Other Methods.
    func percolateDown(_ parent : Int) {
        let lChild : Int = 2 * parent + 1
        let rChild : Int = lChild + 1
        var child : Int = -1
        if (lChild < self.size) {
            child = lChild
        }

        if (rChild < self.size && self.comp(arr[lChild]!, arr[rChild]!)) {
            child = rChild
        }

        if (child != -1 && self.comp(arr[parent]!, arr[child]!)) {
            self.arr.swapAt(parent, child)
			self.percolateDown(child)
        }
    }

    func percolateUp(_ child : Int) {
        let parent : Int = (child - 1) / 2
        if (parent >= 0 && self.comp(arr[parent]!, arr[child]!)) {
			self.arr.swapAt(child, parent)
            self.percolateUp(parent)
        }
    }

    public var isEmpty : Bool {
        return (self.size == 0)
    }

    public var length : Int {
        return self.size
    }

    func peek() -> T? {
        if (self.isEmpty) {
            print("Heap empty exception.")
			return nil
        }
        return self.arr[0]
    }

    func add(_ value : T) {
		self.size += 1
		self.arr.append(value)
        self.percolateUp(self.size - 1)
    }

    func remove() -> T? {
        if (self.isEmpty) {
            print("Heap empty exception.")
			return nil
        }
        let value = self.arr[0]
        self.arr[0] = self.arr[self.size - 1]
        self.size -= 1
		self.arr.removeLast()
        self.percolateDown(0)
        return value
    }

    func display() {
		var i : Int = 0
		while (i < self.size) {
			print(self.arr[i]!, terminator: " ")
			i += 1
		}
        print()
    }

    func delete(_ value : T) -> Bool {
		var i : Int = 0
		while (i < self.size) {
			if (self.arr[i] == value) {
				self.arr[i] = self.arr[self.size - 1]
				self.size -= 1
				self.arr.removeLast()
				self.percolateUp(i)
				self.percolateDown(i)
				return true
			}
			i += 1
		}
        return false
    }
}

class GraphAM {
    var count : Int
    var adj : [[Int]]

    init(_ cnt : Int) {
        self.count = cnt
        self.adj = Array(repeating: Array(repeating: 0, count: self.count), count: self.count)
    }

    func addDirectedEdge(_ src : Int, _ dst : Int, _ cost : Int = 1) {
        self.adj[src][dst] = cost
    }

    func addUndirectedEdge(_ src : Int, _ dst : Int, _ cost : Int = 1) {
        self.addDirectedEdge(src,dst,cost)
        self.addDirectedEdge(dst,src,cost)
    }

    func printGraph() {
        var i : Int = 0
		while (i < self.count) {
			print("Vertex \(i) is connected to : ",terminator: "")
			var j : Int = 0
			while (j < self.count) {
				if (self.adj[i][j] != 0) {
					print("\(j) (cost: \(self.adj[i][j]))",terminator: " ")
				}
				j += 1
			}
			print()
			i += 1
        }
    }

	class Edge : Comparable {
        var src : Int
        var dest : Int
        var cost : Int

        init(_ s : Int, _ d : Int, _ c : Int) {
            self.src = s
            self.dest = d
            self.cost = c
        }

		static func < (lhs: Edge, rhs: Edge) -> Bool {
            return lhs.cost < rhs.cost
        }

		static func > (lhs: Edge, rhs: Edge) -> Bool {
            return lhs.cost > rhs.cost
        }

        static func == (lhs: Edge, rhs: Edge) -> Bool {
            return lhs.cost == rhs.cost
        }
    }

    func dijkstra(_ source : Int) {
        var previous : [Int] = Array(repeating: -1, count: self.count)
        var dist : [Int] = Array(repeating: Int.max, count: self.count)
        var visited : [Bool] = Array(repeating: false, count: self.count)

        dist[source] = 0
        previous[source] = source

        let queue = Heap<Edge>(true)
		var node : Edge = Edge(source, source, 0)
        queue.add(node)

        while (queue.isEmpty != true) {
            node = queue.remove()!
            let src : Int = node.dest
            visited[src] = true
			var dest : Int = 0
			while (dest < self.count) {
				let cost : Int = self.adj[src][dest]
				if (cost != 0) {
					let alt : Int = cost + dist[src]
					if (dist[dest] > alt && visited[dest] == false) {
						dist[dest] = alt
						previous[dest] = src
						node = Edge(src, dest, alt)
						queue.add(node)
					}
				}
				dest += 1
			}
        }
        self.printPath( &previous, &dist,self.count,source)
    }

    func printPathUtil(_ previous : inout [Int], _ source : Int, _ dest : Int) -> String {
        var path : String = ""
        if (dest == source) {
            path += String(source)
        } else {
            path += self.printPathUtil( &previous,source,previous[dest])
            path += ("->" + String(dest))
        }
        return path
    }

    func printPath(_ previous : inout [Int], _ dist : inout [Int], _ count : Int, _ source : Int) {
        var output : String = "Shortest Paths: "
        var i : Int = 0
		while (i < count) {
			if (dist[i] == 99999) {
				output += ("(" + String(source) + "->" + String(i) + " @ Unreachable) ")
			} else if (i != previous[i]) {
				output += "("
				output += self.printPathUtil( &previous,source,i)
				output += (" @ " + String(dist[i]) + ") ")
			}
			i += 1
		}
        print(output)
    }

    func primsMST() {
        var previous : [Int] = Array(repeating: -1, count: self.count)
        var dist : [Int] = Array(repeating: Int.max, count: self.count)
        var visited : [Bool] = Array(repeating: false, count: self.count)

        var source : Int = 0
        dist[source] = 0
        previous[source] = source

        let queue = Heap<Edge>(true)
        var node : Edge = Edge(source, source, 0)
        queue.add(node)

        while (queue.isEmpty != true) {
            node = queue.remove()!
            source = node.dest
            visited[source] = true

			var dest : Int = 0
			while (dest < self.count) {
				let cost : Int = self.adj[source][dest]
				if (cost != 0) {
					if (dist[dest] > cost && visited[dest] == false) {
						dist[dest] = cost
						previous[dest] = source
						node = Edge(source, dest, cost)
						queue.add(node)
					}
				}
				dest += 1
			}
        }

        // printing result.
        var sum : Int = 0
        var isMst : Bool = true
        var output : String = "Edges are "
        var i : Int = 0
		while (i < self.count) {
			if (dist[i] == Int.max) {
				output += ("(" + String(i) + ", Unreachable) ")
				isMst = false
			} else if (previous[i] != i) {
				output += ("(" + String(previous[i]) + "->" + String(i) + " @ " + String(dist[i]) + ") ")
				sum += dist[i]
			}
			i += 1
		}

        if (isMst) {
            print(output)
            print("Total MST cost: " + String(sum))
        } else {
            print("Can\'t get a Spanning Tree")
        }
    }


	func hamiltonianPathUtil(_ path : inout [Int], _ pSize : Int, _ added : inout [Int]) -> Bool {
        // Base case full length path is found
        if (pSize == self.count) {
            return true
        }

		var vertex : Int = 0
		while (vertex < self.count) {
			// There is an edge from last element of path and next vertex
			// and the next vertex is not already included in the path.
			if (pSize == 0 || (self.adj[path[pSize - 1]][vertex] == 1 && added[vertex] == 0)) {
				path[pSize] = vertex
				added[vertex] = 1
				if (self.hamiltonianPathUtil( &path,pSize+1, &added)) {
					return true
				}
				added[vertex] = 0
			}
			vertex += 1
		}
        return false
    }

    func hamiltonianPath() -> Bool {
        var path : [Int] = Array(repeating: 0, count: self.count)
        var added : [Int] = Array(repeating: 0, count: self.count)

        if (self.hamiltonianPathUtil( &path,0, &added)) {
            print("Hamiltonian Path found :: ",terminator: "")
			var i : Int = 0
			while (i < self.count) {
				print(" " + String(path[i]),terminator: "")
				i += 1
			}
            print()
            return true
        }
        print("Hamiltonian Path not found")
        return false
    }

    func hamiltonianCycleUtil(_ path : inout [Int], _ pSize : Int, _ added : inout [Int]) -> Bool {
        // Base case full length path is found
        // this last check can be modified to make it a path.
        if (pSize == self.count) {
            if (self.adj[path[pSize - 1]][path[0]] == 1) {
                path[pSize] = path[0]
                return true
            } else {
                return false
            }
        }

		var vertex : Int = 0
		while (vertex < self.count) {
			// there is a path from last element and next vertex
			if (pSize == 0 || (self.adj[path[pSize - 1]][vertex] == 1 && added[vertex] == 0)) {
				path[pSize] = vertex
				added[vertex] = 1
				if (self.hamiltonianCycleUtil( &path,pSize+1, &added)) {
					return true
				}
				added[vertex] = 0
			}
			vertex += 1
		}
        return false
    }
	
    func hamiltonianCycle() -> Bool {
        var path : [Int] = Array(repeating: 0, count: self.count + 1)
        var added : [Int] = Array(repeating: 0, count: self.count)
        if (self.hamiltonianCycleUtil( &path,0, &added)) {
            print("Hamiltonian Cycle found :: ",terminator: "")
			var i : Int = 0
			while (i <= self.count) {
				print(" " + String(path[i]),terminator: "")
				i += 1
			}
            print("")
            return true
        }
        print("Hamiltonian Cycle not found")
        return false
    }
}

// Testing code.
func main1() {
let gph : GraphAM = GraphAM(4)
gph.addUndirectedEdge(0,1)
gph.addUndirectedEdge(0,2)
gph.addUndirectedEdge(1,2)
gph.addUndirectedEdge(2,3)
gph.printGraph()
}

// Vertex 0 is connected to : 1(cost: 1) 2(cost: 1) 
// Vertex 1 is connected to : 0(cost: 1) 2(cost: 1) 
// Vertex 2 is connected to : 0(cost: 1) 1(cost: 1) 3(cost: 1) 
// Vertex 3 is connected to : 2(cost: 1)

// Testing code.
func main2() {
	let gph : GraphAM = GraphAM(9)
	gph.addUndirectedEdge(0,1,4)
	gph.addUndirectedEdge(0,7,8)
	gph.addUndirectedEdge(1,2,8)
	gph.addUndirectedEdge(1,7,11)
	gph.addUndirectedEdge(2,3,7)
	gph.addUndirectedEdge(2,8,2)
	gph.addUndirectedEdge(2,5,4)
	gph.addUndirectedEdge(3,4,9)
	gph.addUndirectedEdge(3,5,14)
	gph.addUndirectedEdge(4,5,10)
	gph.addUndirectedEdge(5,6,2)
	gph.addUndirectedEdge(6,7,1)
	gph.addUndirectedEdge(6,8,6)
	gph.addUndirectedEdge(7,8,7)
	gph.primsMST()
}

// Edges are (0->1 @ 4) (5->2 @ 4) (2->3 @ 7) (3->4 @ 9) (6->5 @ 2) (7->6 @ 1) (0->7 @ 8) (2->8 @ 2) 
// Total MST cost: 37

// Testing code.
func main3() {
	let gph : GraphAM = GraphAM(9)
	gph.addUndirectedEdge(0,1,4)
	gph.addUndirectedEdge(0,7,8)
	gph.addUndirectedEdge(1,2,8)
	gph.addUndirectedEdge(1,7,11)
	gph.addUndirectedEdge(2,3,7)
	gph.addUndirectedEdge(2,8,2)
	gph.addUndirectedEdge(2,5,4)
	gph.addUndirectedEdge(3,4,9)
	gph.addUndirectedEdge(3,5,14)
	gph.addUndirectedEdge(4,5,10)
	gph.addUndirectedEdge(5,6,2)
	gph.addUndirectedEdge(6,7,1)
	gph.addUndirectedEdge(6,8,6)
	gph.addUndirectedEdge(7,8,7)
	gph.dijkstra(0)
}

// Shortest Paths: (0->1 @ 4) (0->1->2 @ 12) (0->1->2->3 @ 19) (0->7->6->5->4 @ 21) (0->7->6->5 @ 11) (0->7->6 @ 9) (0->7 @ 8) (0->1->2->8 @ 14) 

// Testing code.
func main4() {
	let count : Int = 5
	let gph : GraphAM = GraphAM(count)
	let adj : [[Int]] =
	[[0, 1, 0, 1, 0],
	[1, 0, 1, 1, 0],
	[0, 1, 0, 0, 1],
	[1, 1, 0, 0, 1],
	[0, 1, 1, 1, 0]]

	var i : Int = 0
	while (i < count) {
		var j : Int = 0
		while (j < count) {
			if (adj[i][j] == 1) {
				gph.addDirectedEdge(i,j,1)
			}
			j += 1
		}
		i += 1
	}
	print("hamiltonianPath : " + String(gph.hamiltonianPath()))
	
	let gph2 : GraphAM? = GraphAM(count)
	let adj2 : [[Int]] =
	[
	[0, 1, 0, 1, 0],
	[1, 0, 1, 1, 0],
	[0, 1, 0, 0, 1],
	[1, 1, 0, 0, 0],
	[0, 1, 1, 0, 0]]

	i = 0
	while (i < count) {
		var j : Int = 0
		while (j < count) {
			if (adj2[i][j] == 1) {
				gph2!.addDirectedEdge(i,j,1)
			}
			j += 1
		}
		i += 1
	}
	print("hamiltonianPath :  " + String(gph2!.hamiltonianPath()))
}

// Hamiltonian Path found ::  0 1 2 4 3
// hamiltonianPath : true
// Hamiltonian Path found ::  0 3 1 2 4
// hamiltonianPath :  true


// Testing code.
func main5() {
	let count : Int = 5
	let gph : GraphAM? = GraphAM(count)
	let adj : [[Int]] =
	[[0, 1, 0, 1, 0],
	 [1, 0, 1, 1, 0],
	 [0, 1, 0, 0, 1],
	 [1, 1, 0, 0, 1],
	 [0, 1, 1, 1, 0]]

	var i : Int = 0
	while (i < count) {
		var j : Int = 0
		while (j < count) {
			if (adj[i][j] == 1) {
				gph!.addDirectedEdge(i,j,1)
			}
			j += 1
		}
		i += 1
	}

	print("hamiltonianCycle : " + String(gph!.hamiltonianCycle()))
	let gph2 : GraphAM? = GraphAM(count)
	let adj2 : [[Int]] =
	[[0, 1, 0, 1, 0],
	 [1, 0, 1, 1, 0],
	 [0, 1, 0, 0, 1],
	 [1, 1, 0, 0, 0],
	 [0, 1, 1, 0, 0]]

	i = 0
	while (i < count) {
		var j : Int = 0
		while (j < count) {
			if (adj2[i][j] == 1) {
				gph2!.addDirectedEdge(i,j,1)
			}
			j += 1
		}
		i += 1
	}
	print("hamiltonianCycle :  " + String(gph2!.hamiltonianCycle()))
}

// Hamiltonian Cycle found ::  0 1 2 4 3 0
// hamiltonianCycle : true
// Hamiltonian Cycle not found
// hamiltonianCycle :  false


main1()
main2()
main3()
main4()
main5()
